src/datasets/io/json.py

```diff
@@ -63,49 +64,83 @@ def __init__(
+            written = self._write(file_obj=self.path_or_buf, orient=orient, lines=lines, **self.to_json_kwargs)
         return written
 
+    def _batch_json(self, args):
+        offset, orient, lines = args
+
+        batch = query_table(
+            table=self.dataset.data,
+            key=slice(offset, offset + self.batch_size),
+            indices=self.dataset._indices if self.dataset._indices is not None else None,
+        )
+        var = batch.to_pandas().to_json(path_or_buf=None, orient=orient, lines=lines, **self.to_json_kwargs)
+        return var.encode(self.encoding)
+
     def _write(
         self,
         file_obj: BinaryIO,
-        batch_size: int,
-        encoding: str = "utf-8",
-        orient="records",
-        lines=True,
+        orient,
+        lines,
         **to_json_kwargs,
     ) -> int:
         """Writes the pyarrow table as JSON lines to a binary file handle.
 
         Caller is responsible for opening and closing the handle.
         """
         written = 0
-        _ = to_json_kwargs.pop("path_or_buf", None)
-
-        for offset in utils.tqdm(
-            range(0, len(self.dataset), batch_size), unit="ba", disable=bool(logging.get_verbosity() == logging.NOTSET)
-        ):
-            batch = query_table(
-                table=self.dataset.data,
-                key=slice(offset, offset + batch_size),
-                indices=self.dataset._indices if self.dataset._indices is not None else None,
+
+        if self.num_proc is None or self.num_proc == 1:
+            for offset in utils.tqdm(
+                range(0, len(self.dataset), self.batch_size),
+                unit="ba",
+                disable=bool(logging.get_verbosity() == logging.NOTSET),
+                desc="Creating json from Arrow format",
+            ):
+                batch = query_table(
+                    table=self.dataset.data,
+                    key=slice(offset, offset + self.batch_size),
+                    indices=self.dataset._indices if self.dataset._indices is not None else None,
+                )
+                json_str = batch.to_pandas().to_json(path_or_buf=None, orient=orient, lines=lines, **to_json_kwargs)
+                if not json_str.endswith("\n"):
+                    json_str += "\n"
+                written += file_obj.write(json_str.encode(self.encoding))
+        else:
+            pool = mp.Pool(processes=self.num_proc)
+            json_list = list(
+                utils.tqdm(
+                    pool.imap(
+                        self._batch_json,
+                        [(offset, orient, lines) for offset in range(0, len(self.dataset), self.batch_size)],
+                    ),
+                    total=(len(self.dataset) // self.batch_size) + 1,
+                    unit="ba",
+                    disable=bool(logging.get_verbosity() == logging.NOTSET),
+                    desc="Creating json from Arrow format",
+                )
             )
```

Comment:
This brings all the JSON data in memory right ? This may cause issues on big datasets

Reply:
I ran the process again with this repo for `ascent_kb` dataset. Output json is around 4GB in size and memory consumption varied from 2.62-4GB (it also used swapped memory which peaked at 2GB). Once `written = file_obj.writelines(json_list)` starts then swap memory begins reducing

Reply:
If I understand correctly `json_list` is the list of all the batches of JSON data of all the dataset, so it can be really huge.
The batch size doesn't change the size of what's loaded in RAM since it's basically the full dataset (just separated in batches of size batch_size)